File: DataStructuresC++/Tree/IsThisBinaryTree.cpp
#include <stddef.h>
#include <iostream>
/* Hidden stub code will pass a root argument to the function below. Complete the function to solve the challenge. Hint: you may want to write one or more helper functions.  

The Node struct is defined as follows:
	struct Node {
		int data;
		Node* left;
		Node* right;
	}
*/
	struct Node {
		int data;
		Node* left;
		Node* right;
	};
    bool left(Node*root, int data){
        if (root == NULL){
             return true; 
        }
           
        if (root->data >= data){
            return false;
        }
    

        return ((left(root->left,data)) && (left(root->right,data)) && (left(root->left,root->data)) && (right(root->right,root->data ))); 
    }

    bool right(Node*root, int data){
        if (root == NULL){
             return true; 
        }
           
        if (root->data <= data){
            return false;
        }
    

        return ((right(root->right,data)) && (right(root->left, data)) && (right(root->right, root->data)) && (left(root->left, root->data ))); 
    }

	bool checkBST(Node* root) {
		return ((left(root->left, root->data)) && (right(root->right, root->data)));
	}

File: Data Structures Problems(python)/Tree/IsThisBinaryTree.py
""" Node is defined as
class node:
  def __init__(self, data):
      self.data = data
      self.left = None
      self.right = None
"""
def left(root, data):
    if not root:
        return True 
    if root.data >= data:
        return False
    
    return left(root.left, data) and left(root.right, data) and left(root.left, root.data) and right(root.right, root.data )


def right(root, data):
    if not root:
        return True 
    if root.data <= data:
        return False
    
    return right(root.right, data) and right(root.left, data) and right(root.right, root.data) and left(root.left, root.data)

        
        
def check_binary_search_tree_(root):
    return  left(root.left, root.data) and right(root.right, root.data)
    
    